### PR TITLE
Stop globbing of quoted arguments to native commands

### LINKED ISSUE #3931: globbing for native commands is too agressive
Steps to reproduce
------------------

Intuitively globbing should not kick-in inside the single-quoted strings.

```
echo '11:1' | grep '.*:.'
```

Expected behavior
-----------------
Works, output is `11:1`, like in bash.

Actual behavior
---------------

```
Cannot find drive. A drive with the name '.*' does not exist.
At line:1 char:1
+ echo '11:1' | grep '.*:.'
+ ~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (.*:String) [], DriveNotFoundException
    + FullyQualifiedErrorId : DriveNotFound 
```

The error is pretty confusing for a unix user.

Workaround
------------

Escape `*` by a backtick in the regex.

Environment data
----------------

```powershell
> $PSVersionTable

Name                           Value                                                                                      
----                           -----                                                                                      
PSVersion                      6.0.0-beta                                                                                 
PSEdition                      Core                                                                                       
BuildVersion                   3.0.0.0                                                                                    
CLRVersion                                                                                                                
GitCommitId                    v6.0.0-beta.2                                                                              
OS                             Darwin 16.6.0 Darwin Kernel Version 16.6.0: Fri Apr 14 16:21:16 PDT 2017; root:xnu-3789....
Platform                       Unix                                                                                       
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                                    
PSRemotingProtocolVersion      2.3                                                                                        
SerializationVersion           1.1.0.1                                                                                    
WSManStackVersion              3.0  
```

### PR DESCRIPTION
**Pull request by lzybkr**
Fix globbing so that quoted args are not expanded.

Also Fix some minor issues with exceptions being raised when resolving
the path - falling back to no glob.

The bulk of the change is in the first commit, which passes the `Ast` or `null` instead of the `IScriptExtent` when constructing the parameters and arguments that are used by the parameter binder.

Fix #3931,#4971

### CODE REVIEW

**Comment by rkeithhill** [comment]
Must ... have ... this ... fix.  :-)

**Comment by daxian-dbw** [comment]
@lzybkr I was out in training last Friday. Will get this done today.

File: src/System.Management.Automation/engine/NativeCommandParameterBinder.cs [hunk 1]
```diff
@@ -207,105 +227,103 @@ private void appendOneNativeArgument(ExecutionContext context, object obj, char
...
-                                    _arguments.Append(arg);
-                                }
+                    // Expand paths, but only from the file system.
+                    if (paths?.Count > 0 && paths.All(p => p.BaseObject is FileSystemInfo))
+                    {
+                        var sep = "";
+                        foreach (var path in paths)
+                        {
+                            _arguments.Append(sep);
+                            sep = " ";
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/NativeCommandParameterBinder.cs]: Is `sep = " "` really needed here? It seems we just need `var sep = " "` right before the `foreach`

**Review by daxian-dbw — DISMISSED** [review]

File: src/System.Management.Automation/engine/NativeCommandParameterBinder.cs (hunk 1 above)
Comment (lzybkr) [reply src/System.Management.Automation/engine/NativeCommandParameterBinder.cs]: Well, it's more tidy - it avoids putting 2 spaces between the previous token - either the command name or an argument.

Comment (daxian-dbw) [reply src/System.Management.Automation/engine/NativeCommandParameterBinder.cs]: Thanks for clarifying #Closed.

File: src/System.Management.Automation/engine/NativeCommandParameterBinder.cs [hunk 2]
```diff
@@ -207,105 +227,103 @@ private void appendOneNativeArgument(ExecutionContext context, object obj, char
...
-#else
-                            _arguments.Append(arg);
-#endif
+                            argExpanded = true;
                         }
                     }
                 }
-            } while (list != null);
+            }
+            else
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/NativeCommandParameterBinder.cs]: Now we will expand `~` even if `arg` is not a bare word. Do we want to expand `~` unconditionally?

**Review by daxian-dbw — COMMENTED** [review]
LGTM, but have one more comment about expanding `~`.

File: src/System.Management.Automation/engine/NativeCommandParameterBinder.cs (hunk 2 above)
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/NativeCommandParameterBinder.cs]: Another question: it seems to me that after this change we don't glob for variables. Is it intentional?
For example, before the change
```
PS> $var = '*.txt'
PS> /bin/echo $var                                                                               
a.txt b.txt c.txt
```
with the change
```
PS> /bin/echo $var                                                                                
*.txt
```

Comment (lzybkr) [reply src/System.Management.Automation/engine/NativeCommandParameterBinder.cs]: Thanks for catching, I missed this when reviewing Bruce's changes, but you're right, we don't want to expand `~` unconditionally. I'll fix it.

Comment (lzybkr) [reply src/System.Management.Automation/engine/NativeCommandParameterBinder.cs]: And I see bash also expands variables unless quoted, so we should probably match that behavior too.

Comment (lzybkr) [reply src/System.Management.Automation/engine/NativeCommandParameterBinder.cs]: @daxian-dbw - updated. If you sign off, I'll merge - I want to squash the last 2 commits and rebase 2 commits instead squashing the 3 commits.

**Review by daxian-dbw — APPROVED** [review]
LGTM. @lzybkr please feel free to merge the PR.

**Comment by vors** [comment]
🎉

**Pull request merged**